streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBVersionedStoreSegmentValueFormatter.java

```diff
@@ -495,6 +501,41 @@ private boolean isLastIndex(final int index) {
             return unpackedReversedTimestampAndValueSizes.get(index).timestamp == minTimestamp;
         }
 
+        private void truncateRecordsToTimestamp(final long timestamp) {
+            if (timestamp <= minTimestamp) {
+                // delete everything in this current segment by replacing it with a degenerate segment
+                initializeWithRecord(new ValueAndValueSize(null), timestamp, timestamp);
+                return;
+            }
+
+            final SegmentSearchResult searchResult = find(timestamp, false);
+            // all records with later timestamps should be removed
+            int fullRecordsToTruncate = searchResult.index();
```

Comment:
Don't we know, that we only need to remove the oldest entry at index 0? For what case would we need to remove more than one entry/version ?

Reply:
See above. I also can't think of situations in which we'd need to recovery by doing anything other than removing the current last record version in full, but I think it is good to be defensive. The current understanding that this is the only recovery scenario which is needed is based on a lot of factors (the fact that this recovery is only performed on `insertAsLatest()`, usage of `insertAsLatest()` from the store implementation, etc) which could potentially change in the future. I think it'd be good to future-proof this recovery logic in the event that something changes in the future, even though we do not expect that to be the case.

Reply:
> but I think it is good to be defensive

In general yes, but only if we are sure we don't corrupt data. Are we sure about this?

Reply:
As the code stands today, it's safe yeah. The safety relies on some properties of the current usage of this class, though, such as the fact that `insertAsLatest()` is only called in certain cases, that we always insert into the older segment first when two segments must both be updated, etc. If these change in the future then it could become unsafe. So I guess the "future-proofing" that I intended with these more generic updates creates its own risk for un-future-proof changes as well. 

Maybe the question then is what we think is worse -- if we don't allow generic truncation but instead require that only one full record can be truncated (which is what we expect to happen if there is a failure and then the same record(s) are re-processed right afterward), then the store will throw IllegalStateException when the assumption is violated. Users will have to delete local instances of the store and restore from changelog to recover. Alternatively, if we keep the generic changes, then that is not necessary (and is safe for now), but something could change in the future to make it unsafe (unlikely but possible).

I see the merits of both. Maybe it's better to start with the stricter requirement (do not support generic truncation) first and relax the requirement if users report hitting IllegalStateException. WDYT?

Reply:
Agreed; it's a tricky question. It seems like either way, we need to clearly document/enforce the contract so that future generations can have a reasonably good chance of not breaking the system.

Ideally, there would be at one test verifying the contract is followed (eg, the order of inserts into the underlying segments), so that when someone breaks it, they'll also break the test, which can have a name and comment explaining the contract. The next-best thing would be to structure the code in a way that's pretty self-explanatory, though a sequence of refactors can destroy this property without any specific step being obviously wrong.

I don't think I like the idea of just throwing an IllegalStateException in production code. It puts the user in a really bad position where they've upgraded to a new version that is now crashing in production, and they have to decide whether it's safe to downgrade or whether they need to wait for a fix, and all the while they're incurring downtime. Note that the IllegalState condition may take a while to manifest; imagine what you'd be going through if your app started crashing with this error a month after your last upgrade.

OTOH, we do have recoverable exceptions in Streams, where you can signal to the runtime that the task is corrupted, triggering the purge-and-rebuild-from-changelog automatically. Perhaps that, coupled with an ERROR log telling the user to file a bug ticket, would be the best path forward.

Reply:
Thanks for the discussion, everyone! I chatted with Matthias and we've decided to not throw IllegalStateException or a recoverable exception, and instead log an extra warn message if more (or less) than a single record is truncated as part of this recovery logic, which would indicate either nondeterministic processing or that something has changed in the code to violate our current understanding that only one record should be truncated for recovery. 

As Guozhang noted it's quite difficult to write tests to ensure that all the various existing properties that guarantee safety of this truncation today are preserved going forward as there are a bunch of them which all mutually interact. We have pretty thorough unit test coverage on the outer classes already. The only extra testing I can think to add would be tests which explicitly simulate partial failed updates and verify recovery. Might be overkill but I can do that in a follow-up PR if we think it's worth it.